Add timeout option to catch infinite-loop/long-running tests

justingrant opened the pull request:
Adds a `timeoutMsecs` option to allow tests to run before they'll be terminated. This ensures that infinite-loop (or super-long) tests won't prevent others from completing. Default is 2000 msecs (2 seconds) which should be fine even for slow CI systems. But when running tests in a debugger, set the timeout to much longer (like 1 hour) so that you'll have time to debug tests. If a string is provided, it'll be parsed into a number before evaluation, which makes it easier for callers to pass environment variables as-is. NaN values will silently be assigned the default value.

FYI, this option was added to support using this Test262 runner package with https://github.com/fullcalendar/temporal.

12wrigja commented:
Not a `timeoutDuration` ? :P

Seems reasonable, but doesn't that indicate a bug somewhere in that implementation? Is it that the tests are just slow but eventually complete, or are they infinite-looping?

12wrigja commented on index.mjs:
````diff
@@ -57,6 +57,15 @@ import yaml from 'js-yaml';
  *   built-ins/Temporal/Duration/compare/argument-string-negative-fractional-units.js
  *   built-ins/Temporal/Duration/from/argument-string-negative-fractional-units.js
  *   ```
+ * @property {number|string=2000} timeoutMsecs Optional number of milliseconds
````
? I don't think this is needed / is duplicate with the block in the options bag?

justingrant replied:
Sadly, it's needed. If I remember correctly, VS Code doesn't expand JSDoc typedefs. So if you want to show JSDoc info on the function name, the parameter list, and for individual variables in the options object, then you need to clone in both places. Let me know if you want me to dig up the VS Code bug report asking for this to be changed.

12wrigja replied:
No - that's ok. I didn't know that.

Maybe in the future we should migrate this repo to TS to maybe avoid this problem? :) It might also be nice to vend some typings for this.

justingrant replied:
PRs welcome! :-) I'm not sure porting the implementation to TS would justify the added complexity for one small file of implementation code. But adding a .d.ts could be nice.

justingrant commented:
> Not a `timeoutDuration` ? :P

The main challenge with `timeout*` names is that it's often unclear whether it's msecs. We could make it even longer and say `eachTestTimeoutMsecs` but I think shorter is OK. Holler if not OK with you!

> Seems reasonable, but doesn't that indicate a bug somewhere in that implementation? Is it that the tests are just slow but eventually complete, or are they infinite-looping?

Either one. If a polyfill takes an hour to complete or if it never completes, the impact on users, servers, and PRs is kinda the same: really bad!  Polyfills should have ways to prevent DoS cases as well as infinite loops. 

If I remember correctly, the specific case that this broke in  https://github.com/fullcalendar/temporal (which prompted this new option) were tests that passed an infinite value to a call to `add`, and adding 1 to an infinite value is still infinity.

So I think it's OK to consider timed-out tests as failures, esp. because callers can change the timeout values if some polyfills choose to have inefficient algorithms and want to have longer runtimes. Holler if not OK with you!

justingrant commented:
> Not a `timeoutDuration` ? :P

Oh, lol I just understood your joke. Sorry for being thick the first time. Maybe once Temporal lands in Node trunk you could PR the capability to accept a Temporal.Duration! :-)

justingrant committed:
Add timeout option to catch infinite-loop tests

Adds a `timeoutMsecs` option to allow tests to run before they'll be
terminated. This ensures that infinite-loop (or super-long) tests won't
prevent others from completing. Default is 2000 msecs (2 seconds)
which should be fine even for slow CI systems. But when running
tests in a debugger, set the timeout to much longer (like 1 hour) so
that you'll have time to debug tests. If a string is provided, it'll be
parsed into a number before evaluation, which makes it easier for
callers to pass environment variables as-is. NaN values will silently
be assigned the default value.

FYI, this option was added to support using this Test262 runner
package with https://github.com/fullcalendar/temporal.

12wrigja commented:
Phew :) Glad you got the joke! I'm looking forward to the days where we can represent Durations in a more structured manner. The current impl is fine by me.

Tests timing out are absolutely failures - I was just curious what testcases were causing this only because I was wondering if the js-temporal polyfill impl was running into the same problems.
